### ISSUE TITLE
`libvorbisfile` isn't linked against on Fedora when `libvorbis-devel` is installed

### ISSUE THREAD

**Issue by nasso** [xmake-io/xmake-repo#685]
### Describe the bug
Using the `libvorbis` package when `libvorbis-devel` is installed with DNF causes a link error.

**main.c**
```c
#include <vorbis/codec.h>
#include <vorbis/vorbisfile.h>

int main() {
  OggVorbis_File vf;

  if (ov_fopen("hello.ogg", &vf)) {
    ov_clear(&vf);
  }

  return 0;
}
```

**xmake.lua**
```lua
add_requires("libvorbis ^1.3.7", {configs={with_vorbisenc=false}})

target("test")
  set_kind("binary")
  set_languages("c11")
  set_warnings("all")
  add_files("**.c")
  add_packages("libvorbis")
```

With the two files above in the current working directory, reproduce the bug using docker:

```sh
docker run --rm -itv "$(pwd):/code" fedora:34

# run the following commands in the container
cd code
dnf -y install libvorbis-devel # install libvorbis-devel
bash <(curl -fsSL https://xmake.io/shget.text) # install xmake
xmake --root
```

### Expected behavior
The program links without any error.

### Error output
```
[ 50%]: linking.release test
/usr/bin/gcc -o build/linux/x86_64/release/test build/.objs/test/linux/x86_64/release/main.c.o -m64 -L/usr/lib64 -lvorbis -logg
error: /usr/bin/ld: build/.objs/test/linux/x86_64/release/main.c.o: in function `main':
main.c:(.text+0x54): undefined reference to `ov_fopen'
/usr/bin/ld: main.c:(.text+0x67): undefined reference to `ov_clear'
collect2: error: ld returned 1 exit status
```

> Notice how only `-lvorbis` and `-logg` are passed to the linker.

### Related Environment

Please provide compiling and running environment information:

- xmake version: v2.5.9+202111112157
- os: Fedora 34
- target platform: Fedora 34

### Additional context
On Fedora, `libvorbis` **from DNF** comes as 3 separate libraries: `libvorbis.so`, `libvorbisfile.so` and `libvorbisenc.so`. When trying to use [`libvorbis`](https://github.com/xmake-io/xmake-repo/blob/master/packages/l/libvorbis/xmake.lua) **from xrepo**, the DNF package is detected and xmake does not download it again. However, xmake forgets to pass `-lvorbisfile` to the linker, resulting in a link error.

Here's what happens when `libvorbis-devel` is **NOT** installed on the system:

```
[ 50%]: linking.release test
/usr/bin/g++ -o build/linux/x86_64/release/test build/.objs/test/linux/x86_64/release/main.cpp.o -m64 -L/root/.xmake/packages/l/libvorbis/1.3.7/0c3b1284ea2c4ed898e74501ea460787/lib -L/usr/lib64 -lvorbisenc -lvorbisfile -lvorbis -logg
[100%]: build ok!
```

Notice how both `-lvorbisfile` and `-lvorbisenc` are present, when they were missing in the output of the example above.

**Comment by SirLynix** [comment]
I don't understand, why do you set `with_vorbisenc` to false if you use ov_fopen?

**Comment by nasso** [comment]
> I don't understand, why do you set `with_vorbisenc` to false if you use ov_fopen?

`ov_fopen` is part of `vorbisfile`, not `vorbisenc`

**Comment by SirLynix** [comment]
Right. I think it's because of this:
https://github.com/xmake-io/xmake-repo/blob/master/packages/l/libvorbis/xmake.lua#L41

If on_fetch returns nil, then xmake will try the classical way which will indeed only find libvorbis. This is a problem, I'm not so sure about the best way to fix that.

**Comment by waruqi** [comment]
vorbisfile and vorbisenc are not found by find_package

Can you try run the following commands and let me see the output

```console
xmake l find_package vorbisenc
xmake l find_package vorbisfile
xmake l find_package vorbis
```

> add_requires("libvorbis ^1.3.7", {configs={with_vorbisenc=false}})

BTW, it should be `{vorbisenc = false}`

https://github.com/xmake-io/xmake-repo/blob/1c759cf020e6e7128482f11f11a4876270b610c5/packages/l/libvorbis/xmake.lua#L13-L14

**Comment by waruqi** [comment]
This is because on_fetch did not return the version number, but you forced the specified package version in `add_requires("libvorbis ^1.3.7")`, resulting in a mismatch in the package version returned by on_fetch.

Then, xmake can only degenerate to look up vorbis, vorbisenc and other sub-libraries one by one, so some packages are missing.

I just improved the package:find_package of xmake to return version information, and improved the on_fetch of the xmake-repo/libvorbis package to return version information.

In this way, as long as the version matches, the complete package information can be obtained.

```console
xmake update master
xmake repo -u
xmake f -c
xmake
```

**Comment by SirLynix** [comment]
I still think there should be a way for on_fetch to block/completly override xmake default behaviour in this case

**Comment by nasso** [comment]
> ```
> xmake l find_package vorbisenc
> xmake l find_package vorbisfile
> xmake l find_package vorbis
> ```

gives me:

```
[root@ec7510f35fa9 /]# xmake l find_package vorbisenc
{
  linkdirs = {
    "/usr/lib64"
  },
  links = {
    "vorbisenc"
  }
}
[root@ec7510f35fa9 /]# xmake l find_package vorbisfile
{
  linkdirs = {
    "/usr/lib64"
  },
  links = {
    "vorbisfile"
  }
}
[root@ec7510f35fa9 /]# xmake l find_package vorbis
{
  linkdirs = {
    "/usr/lib64"
  },
  links = {
    "vorbis"
  }
}
```

i tried updating to master and got this when running `xmake f -c`:

```
checking for platform ... linux
checking for architecture ... x86_64
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> gtest 1.11.0 [main:y, gmock:y]
  -> glm 0.9.9+8
  -> bullet3 3.09
  -> nlohmann_json v3.10.0 [from:fx-gltf]
error: bad argument #2 to 'sub' (number has no integer representation)
```

**Comment by waruqi** [comment]
xmake f -c -vD nd let me see the backtrace

**Comment by nasso** [comment]
Here's the backtrace of the error:

```
error: @programdir/core/main.lua:280: @programdir/modules/private/action/require/impl/package.lua:964: bad argument #2 to 'sub' (number has no integer representation)
stack traceback:
    [C]: in function 'sub'
    [@programdir/modules/private/action/require/impl/package.lua:964]: in function 'get_configs_str'
    [...modules/private/action/require/impl/install_packages.lua:219]: in function 'description'
    [@programdir/core/base/utils.lua:332]: in function 'confirm'
    [...modules/private/action/require/impl/install_packages.lua:180]: in function '_get_confirm'
    [...modules/private/action/require/impl/install_packages.lua:580]:
    [@programdir/modules/private/action/require/install.lua:85]:
    [@programdir/actions/config/main.lua:408]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/core/base/task.lua:519]: in function 'run'
    [@programdir/core/main.lua:278]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:371]:

stack traceback:
        [C]: in function 'error'
        @programdir/core/base/os.lua:829: in function 'os.raiselevel'
        (...tail calls...)
        @programdir/core/main.lua:280: in upvalue 'cotask'
        @programdir/core/base/scheduler.lua:371: in function <@programdir/core/base/scheduler.lua:368>
```

**Comment by waruqi** [comment]
I have fixed it, try it again.

xmake update -s dev

**Comment by nasso** [comment]
That worked! Thank you for the quick fix!

**Comment by nasso** [comment]
Will this fix be part of the next xmake release (2.5.10 i presume)? Is it available on `master` or only on the `dev` branch?

**Comment by waruqi** [comment]
> Will this fix be part of the next xmake release (2.5.10 i presume)?

right, but next relase is 2.6.1
